### ISSUE TITLE
输出权重为什么不是二值化的

### ISSUE THREAD

**Issue by feifanzui** [liuzechun/Bi-Real-net#4]
我用python输出了训练得到的model，为什么权重都是小数，不是+1 -1 啊

**Comment by Lovegood-1** [comment]
> You need to set the weights to -1 and 1 in the convolution layers and also freeze the learning rate. Then you retrain the BatchNorm layer to absorb the scaling factor.

excuse me, is what you said had been included in your code?

**Comment by lihengxia** [comment]
> You need to set the weights to -1 and 1 in the convolution layers and also freeze the learning rate. Then you retrain the BatchNorm layer to absorb the scaling factor.

How to  constraint the weights to -1 and 1?And, how to  retrain the BatchNorm layer to absorb the scaling factor?

**Comment by SunzeY** [comment]
the torch model only simulate the bindary caculation process. the saved weigt is stil floating-point.
```python
    def __init__(self, in_chn, out_chn, kernel_size=3, stride=1, padding=1):
        super(HardBinaryConv, self).__init__()
        self.stride = stride
        self.padding = padding
        self.number_of_weights = in_chn * out_chn * kernel_size * kernel_size
        self.shape = (out_chn, in_chn, kernel_size, kernel_size)
        self.weights = nn.Parameter(torch.rand((self.number_of_weights,1)) * 0.001, requires_grad=True)

    def forward(self, x):
        real_weights = self.weights.view(self.shape)
        scaling_factor = torch.mean(torch.mean(torch.mean(abs(real_weights),dim=3,keepdim=True),dim=2,keepdim=True),dim=1,keepdim=True)
        #print(scaling_factor, flush=True)
        scaling_factor = scaling_factor.detach()
        binary_weights_no_grad = scaling_factor * torch.sign(real_weights)
        cliped_weights = torch.clamp(real_weights, -1.0, 1.0)
        binary_weights = binary_weights_no_grad.detach() - cliped_weights.detach() + cliped_weights
        #print(binary_weights, flush=True)
        y = F.conv2d(x, binary_weights, stride=self.stride, padding=self.padding)
```
so there is no problem if you just try some experiments. but if you want to convert the model into true BNN and deploy it onto specific hardware, then what you have to do is retrain the BatchNorm layer, just convert the weigth using sign function
```python
self.weight = sign(self.weight)
```
but I suspect whether it is needed to retrain the BatchNorm layer to absorb the scaling factor.
